stempy/image.cpp

```diff
@@ -131,28 +179,43 @@ STEMImage createSTEMImage(std::vector<Block>& blocks, int rows, int columns,  in
   auto dark = vtkm::cont::make_ArrayHandle(darkFieldMask, numberOfPixels);
 #endif
 
-  for(const Block &block: blocks) {
-    auto data = block.data.get();
-#ifdef VTKm
-    // Transfer the entire block of data at once.
-    auto dataHandle = vtkm::cont::make_ArrayHandle(
-      data, numberOfPixels * block.header.imagesInBlock);
-#endif
-    for(int i=0; i<block.header.imagesInBlock; i++) {
+  // Run the calculations in a thread pool while the data is read from
+  // the disk in the main thread.
+  int numThreads = 4;
+  ThreadPool pool(numThreads);
+
+  // Populate the worker pool
+  vector<future<void>> futures;
+  for (; first != last; ++first) {
+    Block b;
 #ifdef VTKm
-      // Use view to the array already transfered
-      auto view = vtkm::cont::make_ArrayHandleView(
-        dataHandle, i * numberOfPixels, numberOfPixels);
-      auto stemValues = calculateSTEMValuesParallel(view, bright, dark);
+    // Instead of calling _runCalculateSTEMValues directly, we use a
+    // lambda so that we can explicity delete the block. Otherwise,
+    // the block will not be deleted until the threads are destroyed.
+    futures.emplace_back(pool.enqueue([b{ std::move(*first) }, numberOfPixels,
```

Comment:
This doesn't compile on CUDA by the way. It complains about auto type deduction for `b{ std::move(*first) }`. I'm gonna look into fixing it.